### PR TITLE
Activity Feed: comment auto suggestions

### PR DESCRIPTION
**Pull request by martastain**
Implements server-side resolving for entity autosuggestion in comments

### CODE REVIEW

**Comment by Innders** [comment]
Is it too crazy to accept `entityIds` (plural)? You can select more than one entity of the same type at one time.

**Comment by martastain** [comment]
it is crazy... you are always commenting on one entity.

**Comment by Innders** [comment]
> you are always commenting on one entity

Actually you can comment on multiple 😬 Okay let's leave it for now...

File: ayon_server/suggestions/models.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+class ProductSuggestionItem(SuggestionItem):
+    id: str = Field(..., example="af3e4b3e-1b1b-4b3b-8b3b-3b3b3b3b3b3b")
+    name: str = Field(..., example="model_main")
+    product_type: str = Field(..., example="Model")
+    parent: FolderSuggestionItem | None = Field(None)
+
+
+class VersionSuggestionItem(SuggestionItem):
+    id: str = Field(..., example="af3e4b3e-1b1b-4b3b-8b3b-3b3b3b3b3b3b")
+    version: int = Field(..., example=1)
```
Comment (Innders) [inline ayon_server/suggestions/models.py:45]: Could we also get `name` field here please

**Review by Innders — CHANGES_REQUESTED** [review]

**Review by Innders — CHANGES_REQUESTED** [review]
Users seems to be broken
https://github.com/ynput/ayon-backend/assets/49156310/50359cd1-77a6-4468-ae7f-175d889e65cd

File: ayon_server/suggestions/models.py (hunk 1 above)
Comment (martastain) [reply ayon_server/suggestions/models.py:47]: added

**Comment by martastain** [comment]
> Users seems to be broken
> 
>  Screen.Recording.2024-06-04.at.16.34.30.mov

Are those users assigned to anything?

**Comment by Innders** [comment]
> Are those users assigned to anything?

Looks like Demo User is the only user assigned to other tasks in the project. All the other users are not assigned to anything. 

If I add `hotel` user to any task on the project it will now appear in the search results.

I also think **me** should appear as well, I might want to tag myself.

**Comment by Innders** [comment]
Status could influence relevance as well.

Status states ranked by most relevant:

1. In Progress
2. Blocked
3. Not started
4. Done

**Comment by martastain** [comment]
> Status could influence relevance as well.
> 
> Status states ranked by most relevant:
> 
> 1. In Progress
> 2. Blocked
> 3. Not started
> 4. Done
> 

> I cannot use statuses in the query logic as they are configurable and may differ :(

**Comment by Innders** [comment]
> > I cannot use statuses in the query logic as they are configurable and may differ :(

But status **states** do not change? 
image

**Comment by martastain** [comment]
> > > I cannot use statuses in the query logic as they are configurable and may differ :(
> 
> But status **states** do not change? <img alt="image" width="1113" src="https://private-user-

That's true. Okay - it's doable, but I'm not convinced if it's worth another two subqueries. Let's keep that open as a "future extension" :)

File: ayon_server/suggestions/folder.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            u.attrib->>'fullName' as label,
+            r.rel_count as has_task
+        FROM users u
+        LEFT JOIN LATERAL (
+            SELECT count(*) as rel_count
+            FROM project_{project_name}.tasks t
+            JOIN project_{project_name}.hierarchy h ON t.folder_id = h.id
+            WHERE
+                t.assignees @> ARRAY[u.name]
+                AND h.path LIKE '{folder.path.lstrip("/")}%'
```
Comment (BigRoy) [inline ayon_server/suggestions/folder.py:48]: Does passing these strings in directly like this into the query mean we're now prone to SQL injection attacks?

Comment (martastain) [reply ayon_server/suggestions/folder.py:48]: project name is sanitized at the API level (in `ayon_server.api.dependencies`) and it won't allow any potentially malicious characters.

folder path is auto-generated as a materialized view from folder names, which also have strict rules on what characters could be used. Without handling queries as strings, I'd be f**** at many different places :)

Postgres does not allow passing table names as query arguments, so specifying schema as part of the query is necessary for all project-scoped FROM clauses.

**Comment by Innders** [comment]
> That's true. Okay - it's doable, but I'm not convinced if it's worth another two subqueries. Let's keep that open as a "future extension" :)

Yeah cool, just throwing ideas around 👍

**Pull request merged**